### PR TITLE
feat(temple): Implement Generator power range calculator

### LINKED ISSUE #165: feat(temple): Implement Generator power range calculator
## Summary

Implement a power range calculator that determines which rooms are powered by Generator rooms based on tier-dependent range.

## Parent Epic
- #161 (Atziri Temple Optimizer)

## Generator Power Mechanics

```
GENERATOR POWER RANGE BY TIER:
┌─────────────────────────────────────────────────────────────┐
│ Tier 1 (Dynamo):              Range = 3 tiles              │
│ Tier 2 (Shrine of Empowerment): Range = 4 tiles           │
│ Tier 3 (Solar Nexus):          Range = 5 tiles            │
└─────────────────────────────────────────────────────────────┘

POWER SPREAD (Manhattan distance from Generator):
┌───┬───┬───┬───┬───┬───┬───┬───┬───┐
│   │   │   │ 3 │   │   │   │   │   │
├───┼───┼───┼───┼───┼───┼───┼───┼───┤
│   │   │ 3 │ 2 │ 3 │   │   │   │   │
├───┼───┼───┼───┼───┼───┼───┼───┼───┤
│   │ 3 │ 2 │ 1 │ 2 │ 3 │   │   │   │
├───┼───┼───┼───┼───┼───┼───┼───┼───┤
│ 3 │ 2 │ 1 │ G │ 1 │ 2 │ 3 │   │   │  G = Generator
├───┼───┼───┼───┼───┼───┼───┼───┼───┤
│   │ 3 │ 2 │ 1 │ 2 │ 3 │   │   │   │  Numbers = distance
├───┼───┼───┼───┼───┼───┼───┼───┼───┤
│   │   │ 3 │ 2 │ 3 │   │   │   │   │
├───┼───┼───┼───┼───┼───┼───┼───┼───┤
│   │   │   │ 3 │   │   │   │   │   │
└───┴───┴───┴───┴───┴───┴───┴───┴───┘
```

## Implementation

```typescript
// src/engine/temple/PowerRangeCalculator.ts

export class PowerRangeCalculator {
  /**
   * Calculate which cells are powered by generators.
   * Returns a map of position -> power sources.
   */
  calculatePowerCoverage(grid: TempleGrid): PowerMap {
    const powerMap = new PowerMap();
    
    // Find all generators
    const generators = this.findGenerators(grid);
    
    for (const gen of generators) {
      const range = this.getRange(gen.tier);
      const poweredCells = this.getCellsInRange(grid, gen.x, gen.y, range);
      
      for (const cell of poweredCells) {
        powerMap.addPowerSource(cell.x, cell.y, {
          generatorPosition: { x: gen.x, y: gen.y },
          generatorTier: gen.tier,
          distance: this.manhattanDistance(gen.x, gen.y, cell.x, cell.y)
        });
      }
    }
    
    return powerMap;
  }
  
  private getRange(tier: 1 | 2 | 3): number {
    return [3, 4, 5][tier - 1];
  }
  
  private getCellsInRange(
    grid: TempleGrid,
    gx: number,
    gy: number,
    range: number
  ): Position[] {
    const cells: Position[] = [];
    
    for (let dy = -range; dy <= range; dy++) {
      for (let dx = -range; dx <= range; dx++) {
        const x = gx + dx;
        const y = gy + dy;
        
        // Check bounds
        if (x < 0 || x >= 9 || y < 0 || y >= 9) continue;
        
        // Check Manhattan distance
        if (Math.abs(dx) + Math.abs(dy) <= range) {
          cells.push({ x, y });
        }
      }
    }
    
    return cells;
  }
  
  /**
   * Important: Generator requires PATH connection to power rooms.
   * This is unlike synergistic rooms that auto-connect.
   */
  isValidPowerPath(grid: TempleGrid, from: Position, to: Position): boolean {
    // BFS to check if there's a path connection
    return this.hasPathConnection(grid, from, to);
  }
  
  private manhattanDistance(x1: number, y1: number, x2: number, y2: number): number {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
  }
}

class PowerMap {
  private powered: Map<string, PowerSource[]> = new Map();
  
  addPowerSource(x: number, y: number, source: PowerSource): void {
    const key = `${x},${y}`;
    const sources = this.powered.get(key) ?? [];
    sources.push(source);
    this.powered.set(key, sources);
  }
  
  isPowered(x: number, y: number): boolean {
    return this.powered.has(`${x},${y}`);
  }
  
  getPowerSources(x: number, y: number): PowerSource[] {
    return this.powered.get(`${x},${y}`) ?? [];
  }
  
  /**
   * Get visual representation for UI overlay.
   */
  getPowerOverlay(): PowerOverlayData {
    return {
      poweredCells: [...this.powered.keys()].map(k => {
        const [x, y] = k.split(',').map(Number);
        return { x, y, sources: this.powered.get(k)! };
      })
    };
  }
}

interface PowerSource {
  generatorPosition: Position;
  generatorTier: 1 | 2 | 3;
  distance: number;
}
```

## Visual Representation

```
┌─────────────────────────────────────────────────────────────┐
│ GENERATOR POWER OVERLAY (shown when hovering generator)     │
├─────────────────────────────────────────────────────────────┤
│                                                              │
│    [ ]  [ ]  [⚡]  [⚡]  [⚡]  [ ]  [ ]  [ ]  [ ]            │
│    [ ]  [⚡]  [⚡]  [⚡]  [⚡]  [⚡]  [ ]  [ ]  [ ]            │
│    [⚡]  [⚡]  [⚡]  [GEN] [⚡]  [⚡]  [⚡]  [ ]  [ ]           │
│    [ ]  [⚡]  [⚡]  [⚡]  [⚡]  [⚡]  [ ]  [ ]  [ ]            │
│    [ ]  [ ]  [⚡]  [⚡]  [⚡]  [ ]  [ ]  [ ]  [ ]             │
│                                                              │
│ ⚡ = Powered cell (rooms here get "powered" bonus)          │
│ Range shown for T1 Generator (3 tiles)                      │
└─────────────────────────────────────────────────────────────┘
```

## Special Rules

1. **Path Connection Required**: Unlike synergistic rooms, Generator only powers rooms that have a valid path connection to it
2. **Multiple Generators**: A room can be powered by multiple generators (stacking not confirmed)
3. **Generator Powers Itself**: Generator is always considered powered
4. **Tier-Based Range**: Range increases with tier (3/4/5)

## Files to Create

- `src/engine/temple/PowerRangeCalculator.ts`
- `src/engine/temple/PowerMap.ts`

## Dependencies
- #162 (Room database)

## Acceptance Criteria
- [ ] Correctly calculates power range per tier
- [ ] Handles multiple overlapping generators
- [ ] Validates path connections for power
- [ ] Provides overlay data for UI visualization
- [ ] Unit tests for power range scenarios

### PR DESCRIPTION
**Pull request by polaz**
## Summary
- Implement PowerRangeCalculator that determines which rooms are powered by Generator rooms based on tier-dependent range
- Add PowerMap class for tracking power sources, distances, and overlapping coverage
- Validate path connections for power (unlike synergistic rooms that auto-connect)

## Changes
- `src/engine/temple/PowerMap.ts` - Power coverage tracking with source information
- `src/engine/temple/PowerRangeCalculator.ts` - BFS-based power calculation with path validation  
- `src/engine/temple/index.ts` - Export new modules
- `test/unit/engine/temple/PowerRangeCalculator.test.ts` - 44 unit tests

## Key Features
- **Tier-based ranges**: T1=3, T2=4, T3=5 tiles (Manhattan distance)
- **Path validation**: Power only flows through connected paths (BFS)
- **Multiple generators**: Tracks overlapping power sources per cell
- **UI overlay data**: Distance gradients for visualization

## Test plan
- [x] Unit tests for PowerMap class (12 tests)
- [x] Unit tests for PowerRangeCalculator (32 tests)
- [x] Edge cases: empty layout, corner generators, many generators
- [x] All 125 temple tests pass

Closes https://github.com/polaz/pob2-web/issues/165

### CODE REVIEW

File: src/engine/temple/PowerMap.ts [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        x,
+        y,
+        sources: matchingSources,
+        minDistance,
+        isGenerator,
+      });
+    }
+
+    return cells;
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/temple/PowerMap.ts]: Documented. Added @remarks to class JSDoc explaining that query methods return data that should be treated as read-only. For UI binding with Vue reactivity, this is safe as the system works with references without mutation.

File: src/engine/temple/PowerMap.ts [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+   * on trusted data from our own maps.
+   *
+   * @param key - Position key string "x,y" (must be from positionKey())
+   * @returns Position with x and y coordinates
+   */
+  private parsePositionKey(key: string): Position {
+    const parts = key.split(',');
+    return {
+      x: Number(parts[0]),
+      y: Number(parts[1]),
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/temple/PowerMap.ts]: Fixed. Added runtime validation with parseInt (base 10) and Number.isNaN check. Now throws Error on malformed input to fail fast and prevent silent bugs from corrupted internal state.

[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/temple/PowerMap.ts]: Fixed as part of the parsePositionKey validation. Using `parts[0] ?? ''` with parseInt and NaN check to handle malformed input safely. Throws descriptive error on invalid keys.

File: src/engine/temple/PowerRangeCalculator.ts [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+    y2: number
+  ): boolean {
+    return layout.connections.some(
+      (conn) =>
+        (conn.from.x === x1 && conn.from.y === y1 &&
+          conn.to.x === x2 && conn.to.y === y2) ||
+        (conn.from.x === x2 && conn.from.y === y2 &&
+          conn.to.x === x1 && conn.to.y === y1)
+    );
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/temple/PowerRangeCalculator.ts]: Fixed. Implemented the adjacency map optimization:
1. Added buildAdjacencyMap() method that converts connections array to Map<string, Set<string>> for O(1) lookups
2. calculatePowerCoverage now builds the map once and passes it to calculateReachableCells for each generator
3. BFS now uses the adjacency map directly instead of calling hasPathConnection
4. Removed the now-unused hasPathConnection method to avoid dead code
